src/trcb.erl

```diff
@@ -23,11 +23,40 @@
 
 -include("ishikawa.hrl").
 
+-export([causal_delivery/3, try_to_deliever/2]).
+
 %% Broadcast message.
 -callback tcbcast(message()) -> ok.
 
 %% Deliver a message.
--callback tcbdeliver(message(), timestamp()) -> ok.
+-callback tcbdeliver(actor(), message(), timestamp()) -> ok.
 
 %% Determine if a timestamp is stable.
 -callback tcbstable(timestamp()) -> {ok, boolean()}.
+
+%% check if a message should be deliver and deliver it, if not add it to the queue
+-spec causal_delivery({actor(), message(), timestamp()}, timestamp(), [{actor(), message(), timestamp()}]) -> {timestamp(), [{actor(), message(), timestamp()}]}.
+causal_delivery({Origin, Msg, MsgVV}, VV, Queue) ->
+    case vclock:dominates(MsgVV, VV) of
+        true ->
+            NewVV = vclock:increment(Origin, VV),
+            %% TODO: Handle message,
+            try_to_deliever(Queue, {NewVV, Queue});
+        false ->
+            {VV, [Queue | [{Origin, Msg, MsgVV}]]}
+    end.
+
+%% Check for all messages in the queue to be delivered
+%% Called upon delievery of a new message that could affect the delivery of messages in the queue
+-spec try_to_deliever([{actor(), message(), timestamp()}], {timestamp(), [{actor(), message(), timestamp()}]}) -> {timestamp(), [{actor(), message(), timestamp()}]}.
+try_to_deliever([], {VV, Queue}) -> {VV, Queue};
+try_to_deliever([{Origin, MsgVV, _Msg}=El | RQueue], {VV, Queue}=V) ->
+    case vclock:dominates(MsgVV, VV) of
+        true ->
+            NewVV = vclock:increment(Origin, VV),
+            %% TODO: Handle message,
```

Comment:
In the pure op we would be adding to the POLog. In general depends what do you want to do with message. Print it, etc.